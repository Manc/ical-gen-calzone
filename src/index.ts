// import type { ICalComponent } from 'ical-gen';
import { ICalCalendar, ICalTimeZone } from 'ical-gen';
import { getZoneLines } from 'icalzone';

function vTimeZoneGenerator(tz: string): ICalTimeZone | undefined {
	const tzLines = getZoneLines(tz, false);
	return tzLines && new ICalTimeZone({ fromLines: tzLines });
};

export default class ICalCalendarWithZones extends ICalCalendar {
	/**
	 * This implementation of `getChildrenForRender()` resolves all time zones
	 * to time zone components and dynamically adds them in front of all other
	 * other children.
	 */
	public getChildrenForRender = async () => {
		// Get all time zones, but ignore UTC.
		const timeZones = this.getTimeZones().filter(tz => tz !== 'UTC');

		// Try to generate a `VTIMEZONE` string for each time zone.
		const tzComponents: Array<ICalTimeZone> = timeZones
			.map(vTimeZoneGenerator)
			.filter(component => !!component) as ICalTimeZone[];

		return [
			...tzComponents, // Insert `VTIMEZONE`s first
			...this.children, // Then the rest like `VENENT` etc.
		];
	};
}
