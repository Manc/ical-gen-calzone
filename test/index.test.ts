import { ICalEvent } from 'ical-gen';
import { ICalCalendarCalzone, ICalCalendarCalzoneOptions } from '../src';


describe('ICalCalendarCalzone', () => {
	const demoEvent1 = new ICalEvent({
		// Note: For the given dates in time zone Europe/London,
		// the resulting local time should be +1 hour (British
		// Summer Time).
		start: {
			date: new Date('2021-05-15T00:00:00.000Z'),
			zone: 'Europe/London',
		},
		end: {
			date: new Date('2021-05-16T00:00:00.000Z'),
			zone: 'Europe/London',
		},
		sequence: 1,
		stamp: new Date('2021-05-15T00:00:00.000Z'),
		summary: 'Test 1',
		uid: 'testuid1',
	});

	const demoEvent2 = new ICalEvent({
		// Note: For the given dates in time zone Europe/Berlin,
		// the resulting local time should be +2 hour (Central
		// European Summer Time).
		start: {
			date: new Date('2021-05-15T00:00:00.000Z'),
			zone: 'Europe/Berlin',
		},
		end: {
			date: new Date('2021-05-16T00:00:00.000Z'),
			zone: 'Europe/Berlin',
		},
		sequence: 1,
		stamp: new Date('2021-05-15T00:00:00.000Z'),
		summary: 'Test 2',
		uid: 'testuid2',
	});

	describe('renderToString()', () => {
		it('renders the calendar with all three time zone components', async () => {
			// Initialize options object separately to test the TS type.
			const options: ICalCalendarCalzoneOptions = {
				prodId: {
					company: 'My Company X',
					product: 'My Product Name',
					language: 'XX',
				},
				children: [demoEvent1, demoEvent2],
				timezone: 'Europe/Paris',
			};
			const str = await new ICalCalendarCalzone(options).renderToString();
			expect(str.match(/BEGIN:VTIMEZONE/g) || []).toHaveLength(3);
			expect(str).toContain('TZID:Europe/Paris');
			expect(str).toContain('TZID:Europe/London');
			expect(str).toContain('TZID:Europe/Berlin');
			expect(str).toMatchSnapshot();
		});

		it('renders the calendar with resolved time zone components, ingnores the unresolved time zone', async () => {
			const str = await new ICalCalendarCalzone({
				prodId: {
					company: 'My Company X',
					product: 'My Product Name',
					language: 'XX',
				},
				children: [demoEvent1, demoEvent2],
				timezone: 'Something/Wrong',
			}).renderToString();
			expect(str.match(/BEGIN:VTIMEZONE/g) || []).toHaveLength(2);
			expect(str).toContain('TZID:Europe/London');
			expect(str).toContain('TZID:Europe/Berlin');
			expect(str).not.toContain('TZID:Something/Wrong');
			expect(str).toMatchSnapshot();
		});
	});
});
