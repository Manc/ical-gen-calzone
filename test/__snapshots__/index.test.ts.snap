// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ICalCalendarCalzone renderToString() renders the calendar with all three time zone components 1`] = `
"BEGIN:VCALENDAR
PRODID:-//My Company X//My Product Name//XX
VERSION:2.0
CALSCALE:GREGORIAN
TIMEZONE-ID:Europe/Paris
X-WR-TIMEZONE:Europe/Paris
BEGIN:VTIMEZONE
TZID:Europe/Paris
BEGIN:STANDARD
TZNAME:CET
TZOFFSETFROM:+0200
TZOFFSETTO:+0100
DTSTART:19701025T030000
RRULE:FREQ=YEARLY;BYMONTH=10;BYDAY=-1SU
END:STANDARD
BEGIN:DAYLIGHT
TZNAME:CEST
TZOFFSETFROM:+0100
TZOFFSETTO:+0200
DTSTART:19700329T020000
RRULE:FREQ=YEARLY;BYMONTH=3;BYDAY=-1SU
END:DAYLIGHT
END:VTIMEZONE
BEGIN:VTIMEZONE
TZID:Europe/London
BEGIN:STANDARD
TZNAME:GMT
TZOFFSETFROM:+0100
TZOFFSETTO:+0000
DTSTART:19701025T020000
RRULE:FREQ=YEARLY;BYMONTH=10;BYDAY=-1SU
END:STANDARD
BEGIN:DAYLIGHT
TZNAME:BST
TZOFFSETFROM:+0000
TZOFFSETTO:+0100
DTSTART:19700329T010000
RRULE:FREQ=YEARLY;BYMONTH=3;BYDAY=-1SU
END:DAYLIGHT
END:VTIMEZONE
BEGIN:VTIMEZONE
TZID:Europe/Berlin
BEGIN:STANDARD
TZNAME:CET
TZOFFSETFROM:+0200
TZOFFSETTO:+0100
DTSTART:19701025T030000
RRULE:FREQ=YEARLY;BYMONTH=10;BYDAY=-1SU
END:STANDARD
BEGIN:DAYLIGHT
TZNAME:CEST
TZOFFSETFROM:+0100
TZOFFSETTO:+0200
DTSTART:19700329T020000
RRULE:FREQ=YEARLY;BYMONTH=3;BYDAY=-1SU
END:DAYLIGHT
END:VTIMEZONE
BEGIN:VEVENT
UID:testuid1
SEQUENCE:1
DTSTAMP:20210515T000000Z
DTSTART;TZID=Europe/London:20210515T010000
DTEND;TZID=Europe/London:20210516T010000
SUMMARY:Test 1
END:VEVENT
BEGIN:VEVENT
UID:testuid2
SEQUENCE:1
DTSTAMP:20210515T000000Z
DTSTART;TZID=Europe/Berlin:20210515T020000
DTEND;TZID=Europe/Berlin:20210516T020000
SUMMARY:Test 2
END:VEVENT
END:VCALENDAR"
`;

exports[`ICalCalendarCalzone renderToString() renders the calendar with resolved time zone components, ingnores the unresolved time zone 1`] = `
"BEGIN:VCALENDAR
PRODID:-//My Company X//My Product Name//XX
VERSION:2.0
CALSCALE:GREGORIAN
TIMEZONE-ID:Something/Wrong
X-WR-TIMEZONE:Something/Wrong
BEGIN:VTIMEZONE
TZID:Europe/London
BEGIN:STANDARD
TZNAME:GMT
TZOFFSETFROM:+0100
TZOFFSETTO:+0000
DTSTART:19701025T020000
RRULE:FREQ=YEARLY;BYMONTH=10;BYDAY=-1SU
END:STANDARD
BEGIN:DAYLIGHT
TZNAME:BST
TZOFFSETFROM:+0000
TZOFFSETTO:+0100
DTSTART:19700329T010000
RRULE:FREQ=YEARLY;BYMONTH=3;BYDAY=-1SU
END:DAYLIGHT
END:VTIMEZONE
BEGIN:VTIMEZONE
TZID:Europe/Berlin
BEGIN:STANDARD
TZNAME:CET
TZOFFSETFROM:+0200
TZOFFSETTO:+0100
DTSTART:19701025T030000
RRULE:FREQ=YEARLY;BYMONTH=10;BYDAY=-1SU
END:STANDARD
BEGIN:DAYLIGHT
TZNAME:CEST
TZOFFSETFROM:+0100
TZOFFSETTO:+0200
DTSTART:19700329T020000
RRULE:FREQ=YEARLY;BYMONTH=3;BYDAY=-1SU
END:DAYLIGHT
END:VTIMEZONE
BEGIN:VEVENT
UID:testuid1
SEQUENCE:1
DTSTAMP:20210515T000000Z
DTSTART;TZID=Europe/London:20210515T010000
DTEND;TZID=Europe/London:20210516T010000
SUMMARY:Test 1
END:VEVENT
BEGIN:VEVENT
UID:testuid2
SEQUENCE:1
DTSTAMP:20210515T000000Z
DTSTART;TZID=Europe/Berlin:20210515T020000
DTEND;TZID=Europe/Berlin:20210516T020000
SUMMARY:Test 2
END:VEVENT
END:VCALENDAR"
`;
